### ISSUE TITLE
HTTPS

### ISSUE THREAD

**Issue by dsolstad** [CTFd/CTFd#581]
Maybe mention in the documentation that you can run CTFd with HTTPS, with self-signed TLS certificate, by adding ssl_contex="adhoc" as a parameter to the run() method in serve.py.

Crucial when running a CTF contest over wifi.

**Comment by tghosth** [comment]
I'm not sure that a self-signed cert is any better than no cert...

If you wanted to use TLS, I created a repository [here](https://github.com/tghosth/CTFd-docker-deploy) containing a docker compose file (`docker-compose-production.yml`) and various other config files prepared by @nategraf. It includes an nginx container to handle TLS if you supply it with a certificate. I am hoping to write up an explanation in a few days.

**Comment by dsolstad** [comment]
Why do you think self-signed is no better? It doesn't protect against mitm,
but at least the traffic gets encrypted and can't be viewed by anyone with
a promiscuous WiFi adapter.

**Comment by tghosth** [comment]
If someone is taking the trouble to sniff wi-fi traffic, it doesn't take much more effort to spoof a self-signed cert as well

**Comment by dsolstad** [comment]
It's a big leap from passively listening to traffic, to say, capturing
others flag submissions, to actively setting up a mitm proxy and forging a
certificate.

**Comment by nategraf** [comment]
There are a ton of tools out there that make it really easy (e.g. sslstrip, ettercap, bettercap)

What is barring you from getting a cert from Let's Encrypt? (or some other provider)

**Comment by ZetaTwo** [comment]
Yeah, just run it with a reverse proxy in front handling the TLS. No point in bogging the app down with TLS stuff.

**Comment by emonte02** [comment]
+1 on the Let's Encrypt option. would be great to spin up SSL protected scoreboards quickly.

**Comment by mike-bailey** [comment]
Having been in this position, you can't just grab a Lets Encrypt cert (or any really short of buying into an intermediate CA) if the site isn't public facing, i.e. if it's over a LAN.

**Comment by mike-bailey** [comment]
But +1 on reverse proxying, that's how ours is setup at GMU.

**Comment by nategraf** [comment]
Here are three ideas in order of ease for the user:
1. Register a public domain and register a Let's Encrypt certificate for it. Then setup a DNS server or proxy on your LAN and point that public domain to your private ip. Serve up CTFd on the LAN with your new cert and profit
2. Create a CA cert, install it on all of your users' computers and use that cert to create your server cert. Congratulations, you are now in the CA business, don't lose your track of that cert :fearful:
3. Serve your traffic with a self signed cert and add a browser security exception to all of your users browsers then make them pinky-promise not to click through any security warnings from then on.

The only place that 2 or 3 would be appropriate is maybe a school, but even then it should give you the heebie jeebies

**Comment by mike-bailey** [comment]
Fair, more making a point as to why people are using self signed for this. For instance for a decent % of our CTFs we don’t have outbound to internet due to the fact that they don’t want to share a pipe with a bunch of hackers

**Comment by nategraf** [comment]
Right, but HTTPS is fundamentally based on a chain of trust, not just cryptography. You need to establish trust with your users somehow and the issue with self-signed is that trust is established as they click through the browser security warning, at which point they have no way to actually verify that they are talking to the right server. Plus if you tell your users to click through the warning, and then in the middle of the competition it happens again, they will likely click through that one too at which point they may be in a mitm attack. Self-signed has no chain of trust

**Comment by mike-bailey** [comment]
Aware of that, but situations exist in which you can only use a self signed cert.

**Comment by mike-bailey** [comment]
If you trust the certificate  afterwhich it changes, say via an intercept. then you’d logically get a new warning.

**Comment by ColdHeat** [comment]
This is about CTFd using HTTPS not about whether or not self signed is a secure solution. 

Put very simply, my opinion is that CTFd dealing with managing SSL for the user is a bad idea. There's no way it would account for every situation and probably take away from implementing other more interesting features. certbot exists and it can be used in situations where you do not have internet. And people are also capable of buying SSL as archaic as it may seem.

**Comment by mike-bailey** [comment]
You can’t verify trust with the CA without internet, can you....? I don’t want to get into a rabbit hole so I will say I do agree the application shouldn’t be responsible for certificates. That should be left to the server handling your requests regardless of how you choose to implement that.
